### PR TITLE
[Unholy DK] Added Vile Infusion tracker v2

### PR DESCRIPTION
**Pull request by Bicepspump**
### Description

Added a tracker for Vile Infusion uptime.

### Motivation

Upkeeping the unholy dk 2p bonus is important to maximise DPS. Either through passively weaving your abilities or actively managing it, it's important.

### Testing

- Test report URL: https://www.warcraftlogs.com/reports/jXn6BbyPh97JV1Fm/#fight=12&source=58
image

### CODE REVIEW

File: src/common/SPELLS/deathknight.ts [hunk 1]
```diff
@@ -301,6 +301,12 @@ const spells = spellIndexableList({
     icon: 'ability_creature_disease_02',
   },
 
+  VILE_INFUSION_BUFF: {
```
Comment (Khazakdk) [inline src/common/SPELLS/deathknight.ts]:
```suggestion
  T29_VILE_INFUSION_BUFF: {
```
Marking this with T29 so its easier to find and remove later

**Review by Khazakdk — APPROVED** [review]
Looks good!

File: src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  constructor(options: Options) {
+    super(options);
+
+    this.active = this.selectedCombatant.has2PieceByTier(TIERS.T29);
+    if (!this.active) {
+      return;
+    }
+
+    this.addEventListener(
+      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.VILE_INFUSION_BUFF),
```
Comment (Khazakdk) [inline src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx]:
```suggestion
      Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.T29_VILE_INFUSION_BUFF),
```

File: src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      this.onBuffEvent,
+    );
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.VILE_INFUSION_BUFF),
```
Comment (Khazakdk) [inline src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx]:
```suggestion
      Events.removebuff.by(SELECTED_PLAYER).spell(SPELLS.T29_VILE_INFUSION_BUFF),
```

File: src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  suggestions(when: When) {
+    // Vile Infusion should have 75% uptime or more
+    when(this.averageBuffUptime)
+      .isLessThan(0.75)
+      .addSuggestion((suggest, actual, recommended) =>
+        suggest(
+          <span>
+            You are not keeping up your <SpellLink id={SPELLS.VILE_INFUSION_BUFF.id} /> enough.{' '}
```
Comment (Khazakdk) [inline src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx]:
```suggestion
            You are not keeping up your <SpellLink id={SPELLS.T29_VILE_INFUSION_BUFF.id} /> enough.{' '}
```

File: src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  statistic() {
+    return (
+      <Statistic
+        tooltip={`Your Vile Infusion was up ${Math.round(this.totalBuffTime)} out of ${Math.round(
+          this.owner.fightDuration / 1000,
+        )} seconds`}
+        position={STATISTIC_ORDER.CORE(1)}
+        size="flexible"
+      >
+        <BoringSpellValueText spellId={SPELLS.VILE_INFUSION_BUFF.id}>
```
Comment (Khazakdk) [inline src/analysis/retail/deathknight/unholy/modules/spells/VileInfusion.tsx]:
```suggestion
        <BoringSpellValueText spellId={SPELLS.T29_VILE_INFUSION_BUFF.id}>
```

**Comment by Pewtro** [comment]
You can use the Pets class to get pet buff uptimes. 

```ts
  static dependencies = {
    pets: Pets,
  };
  protected pets!: Pets;
  get averageBuffUptime() {
    return this.pets.getBuffUptime(SPELLS.VILE_INFUSION_BUFF.id) / this.owner.fightDuration;
  }
  ```

**Review by Khazakdk — APPROVED** [review]

**Pull request merged**